hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/validation/testvalidatorset1/GeneralValidatorsForTesting.java

```diff
@@ -0,0 +1,233 @@
+import static org.apache.hadoop.ozone.om.request.validation.ValidationCondition.OLDER_CLIENT_REQUESTS;
+import static org.apache.hadoop.ozone.om.request.validation.ValidationCondition.UNCONDITIONAL;
+import static org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Type.CreateKey;
+import static org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Type.CreateVolume;
+import static org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.Type.DeleteKeys;
+
+/**
+ * Some annotated request validator method, and facilities to help check if
+ * validations were properly called from tests where applicable.
+ */
+public final class GeneralValidatorsForTesting {
+
+  /**
+   * As the validators written here does not override any request or response
+   * but throw exceptions for specific tests, a test that wants to directly
+   * use a validator here, has to turn on this boolean, and the method that
+   * the test relies on has to check for this value.
+   *
+   * This is necessary to do not affect other tests that are testing requests
+   * processing, as for some of those tests this package is on the classpath
+   * and therefore the annotated validations are loadede for them.
+   */
+  private static boolean validatorTestsRunning = false;
+
+  public static void startValidatorTest() {
+    validatorTestsRunning = true;
+  }
+
+  public static void finishValidatorTest() {
+    validatorTestsRunning = false;
+  }
+
+  private GeneralValidatorsForTesting() { }
+
+  /**
+   * Interface to easily add listeners that get notified if a certain validator
+   * method defined in this class was called.
+   *
+   * @see TestRequestValidations for more details on how this intercace is
+   *      being used.
+   */
+  @FunctionalInterface
+  public interface ValidationListener {
+    void validationCalled(String calledMethodName);
+  }
+
+  private static List<ValidationListener> listeners = new ArrayList<>();
+
+  public static void addListener(ValidationListener listener) {
+    listeners.add(listener);
+  }
+
+  public static void removeListener(ValidationListener listener) {
+    listeners.remove(listener);
+  }
+
+  private static void fireValidationEvent(String calledMethodName) {
+    listeners.forEach(l -> l.validationCalled(calledMethodName));
+  }
+
+  @RequestFeatureValidator(
+      conditions = { CLUSTER_NEEDS_FINALIZATION },
+      processingPhase = PRE_PROCESS,
+      requestType = CreateKey)
+  public static OMRequest preFinalizePreProcessCreateKeyValidator(
+      OMRequest req, ValidationContext ctx) {
+    fireValidationEvent("preFinalizePreProcessCreateKeyValidator");
+    return req;
+  }
+
+  @RequestFeatureValidator(
+      conditions = { CLUSTER_NEEDS_FINALIZATION },
+      processingPhase = POST_PROCESS,
+      requestType = CreateKey)
+  public static OMResponse preFinalizePostProcessCreateKeyValidator(
```

Comment:
Annotating the validator instead of the request isn't bad, but would it be possible to rework it so the annotation goes on the request, and the validator is passed in as a functional interface to the annotation? Right now request annotations like `@DisallowedUntilLayoutVersion` go on the request itself, so if this set of annotations is not on the request, it may be difficult to figure out what processing each request has. Even if I know where to look as an upgrade dev, I think annotating the request would be more obvious/intuitive for the rest of the community who may need to use this feature.

Reply:
As we discussed in one of our meetings, this is an interesting suggestion, and I haven't put too much though in using the annotation on the requests itself before.

If we have the annotation on a request class, then the request type is defined by the request class, so we can skip that, still we remain with a set of methods for four different cases: prefinalize-preprocess, oldclient-preprocess, prefinalize-postprocess, oldclient-postprocess. These methods have to be declared to be used within the annotation, or somewhere. If we want to introduce a new condition or a new processing phase, we would need to introduce more lists for the new cases either in the annotation or elsewhere. An other option might be a naming convention for the methods, but that is a bit harder to check at compile time, and may be error prone.

Pre and post differentiation can not be skipped as the methods for the two has to have a different signature, so it can not fit within the same functional interface. Conditions I believe has to be separated, if not, that would introduce branching in the code of the validator later when any new validation functionality is added.
An alternative for branching when a new validator is added to add multiple validations to a certain case, but in that case the annotation gets bigger and bigger so it becomes harder to see what is happening there.


I know we have mentioned the two phase write request processing as a potential blocker, but all in all that does not seem to be a relevant difference from the point where this framework is integrated to the request processing flow, still I would say it is more beneficial to have small separate methods that do just one thing to be defined as a validator separately, instead of trying to collect all validators in one place in one/multiple annotations before the class declaration, or leaving it to be a potentially growing single method with growing number of code branches, but if you see a potential way of solving this without the mentioned risks, I am open to discuss, I don't see that this change makes life easier or simpler on the long run unfortunately.